### DISCUSSION TITLE
Historizing

### DISCUSSION THREAD

**Discussion by pfetrial** [FreeOpcUa/opcua-asyncio#375] [General]
Hello all,

I made some tests with a Python server . I found out that the minimal data rate change is around 10 ms . 
Is it possible to start/stop historizing ?  per example : to store 1000 data changes and then stop historizing  then trigger/start again the historizing function  to catch some interesting changes? 
I tried to use the functions :
server.historize_node_data_change([myvar1,myvar2,mysin],period=0, count=100)
server.dehistorize_node_data_change([myvar1,myvar2,mysin])

but after the first stop/start historizing, the server makes an error..
Is there another way to historize on demand?
thanks

**Comment by AndreasHeine** [comment]
Hey,

why is it important to start and stop historizing? 
keep in mind that a dynamic addressspaces are tricky for client applications because they dont know when something changes so they need to constantly check for changes this generates alot overhead on clientside aswell on the network.
in my oppinion its more usefull to constantly historize and if a trigger or event happens i read the history!

**Reply by pfetrial** [reply]
Hi Andreas,
Thanks for your reply.
It is important to stop the historization to avoid the overrule the storage with not important datas. Well, if you have a lot of memory/storage you can,  as you pointed out,  read the history when a special event occurs.
Concerning the problem of the call of start ans stop, It has nothing to do with the client because the error is generated by the server. Am i wrong?

**Reply by AndreasHeine** [reply]
in that case it might be easier with an Array[1000] and not opc ua historizing you can create a variant with one dimension, then you are free to do stuff you like and how you want!

by the way what is the errormessage you mentioned? could you post the stacktrace!

**Reply by pfetrial** [reply]
correct ! but we want to use the facilities of the historization of OPC ua ( display )  and In the case we use a Array of samples , we must write a specific software in the PLC .
Here is the stacktrace of the error :


Endpoints other than open requested but private key and certificate are not set.
Listening on 0.0.0.0:4840
Unknown message received FourByteNodeId(i=12208)
start histo call
stop histo call
start histo call
Error executing method call CallMethodRequest(ObjectId:{self.ObjectId}, MethodId:{self.MethodId}, InputArguments:{self.InputArguments}), an exception was raised:
Traceback (most recent call last):
  File "C:\Python38\lib\site-packages\opcua\server\address_space.py", line 478, in _call
    result = node.call(method.ObjectId, *method.InputArguments)
  File "C:\Python38\lib\site-packages\opcua\common\methods.py", line 64, in wrapper
    result = func(parent, *[arg.Value for arg in args])
  File "C:\pfe\python_Filetransfer\minimalserverWithhistorizing.py", line 21, in methodstart
    server.historize_node_data_change([myvar1,myvar2,mysin],period=0, count=2000)
  File "C:\Python38\lib\site-packages\opcua\server\server.py", line 551, in historize_node_data_change
    self.iserver.enable_history_data_change(node, period, count)
  File "C:\Python38\lib\site-packages\opcua\server\internal_server.py", line 235, in enable_history_data_change
    self.history_manager.historize_data_change(node, period, count)
  File "C:\Python38\lib\site-packages\opcua\server\history.py", line 218, in historize_data_change
    self.storage.new_historized_node(node.nodeid, period, count)
  File "C:\Python38\lib\site-packages\opcua\server\history.py", line 91, in new_historized_node
    raise UaNodeAlreadyHistorizedError(node_id)
opcua.server.history.UaNodeAlreadyHistorizedError: NumericNodeId(ns=2;i=2)

thx

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/f189d87b2ab0ad1264fe6aaf05523d4fe27e3c6f/asyncua/server/internal_server.py#L233

https://github.com/FreeOpcUa/opcua-asyncio/blob/f189d87b2ab0ad1264fe6aaf05523d4fe27e3c6f/asyncua/server/internal_server.py#L242

its part of the `server.iserver`
maybe to achive your target you might try that lowlevelmethods!

![123](https://user-images.githubusercontent.com/56362817/103240618-1e08bd80-4951-11eb-8a07-bd1dabbad4d2.PNG)

**Comment by oroulet** [comment]
Not being able to restart historyzing looks like a bug. Make a minimal example then make a bug report.
 But concerning your use case, I really cannot imagine you really need such a feature. It really sounds like an architecture error, Ua is just a protocol, if you need to reduce data storing, it should not be done in protocol but your backend...

**Reply by AndreasHeine** [reply]
not a bug! stop method is for closing db connections before shutdown. maybe a missing api feature to pause the historization. what is missing is checking if the db file exist

**Reply by oroulet** [reply]
if we close the connection maybe we should be able to reconnect?
Anyway as written above I cannot see any use case that make sense. We should not add features that encourage bad practices

**Reply by AndreasHeine** [reply]
> if we close the connection maybe we should be able to reconnect?

therefore we need to catch some sqlite errors if table already exist for example is missing (i just used our examples) atm we catch all errors ...

**Comment by pfetrial** [comment]
From my point of view,  this use case is helpful , perhaps you don’t fully understand the use case : May I explain this in details:

We have a machine controlled by a internal state machine. In the state “execute” ( in the meaning of ISA88) the machine is producing , in the others states the controlled items are not interesting . In the state execute ( during 180 secs ) it’s highly important to see how the machine is reacting with all the 60 items influencing in the process.

This bunch of 60 vars is the image of the process and can be analyzed after the execution to optimize/debug the process .

May this have a sense?

 

In others stack , I have seen such Api to strop/restart histo..

 

 

 

 

Von: oroulet <notifications@github.com> 
Gesendet: Dienstag, 29. Dezember 2020 09:57
An: FreeOpcUa/opcua-asyncio <opcua-asyncio@noreply.github.com>
Cc: pfetrial <patrick.frey37@sfr.fr>; Author <author@noreply.github.com>
Betreff: Re: [FreeOpcUa/opcua-asyncio] Historizing (#375)

 

if we close the connection maybe we should be able to reconnect?
Anyway as written above I cannot see any use case that make sense. We should not add features that encourage bad practices

**Comment by oroulet** [comment]
OK. so you do not want the history server, but you want to stop historysing one or several nodes? I am still not sure about the use case but we should support that kind of things. Do you use the dehistorize() method? make a minimal code showing your issue

**Comment by pfetrial** [comment]
When you say «history server “ do you mean an external history server? 

And yes I want to stop historizing on several nodes.

No I don’t use dehistorize() , as already written I use enable/disable history_data_change. 

Should I use something else?

 

 

Von: oroulet <notifications@github.com> 
Gesendet: Dienstag, 29. Dezember 2020 11:09
An: FreeOpcUa/opcua-asyncio <opcua-asyncio@noreply.github.com>
Cc: pfetrial <patrick.frey37@sfr.fr>; Author <author@noreply.github.com>
Betreff: Re: [FreeOpcUa/opcua-asyncio] Historizing (#375)

 

OK. so you do not want the history server, but you want to stop historysing one or several nodes? I am still not sure about the use case but we should support that kind of things. Do you use the dehistorize() method? make a minimal code showing your issue

**Comment by pfetrial** [comment]
So internally the server use sqllite to store the history ? 

 

Von: Andreas Heine <notifications@github.com> 
Gesendet: Dienstag, 29. Dezember 2020 11:27
An: FreeOpcUa/opcua-asyncio <opcua-asyncio@noreply.github.com>
Cc: pfetrial <patrick.frey37@sfr.fr>; Author <author@noreply.github.com>
Betreff: Re: [FreeOpcUa/opcua-asyncio] Historizing (#375)

 

therefore we need to catch some sqlite errors if table already exist for example is missing

**Reply by AndreasHeine** [reply]
if you use the default historymanager then yes

**Comment by pfetrial** [comment]
Yes I do. But does it use files for storage ? or does it use ram?

 

 

Von: Andreas Heine <notifications@github.com> 
Gesendet: Dienstag, 29. Dezember 2020 15:47
An: FreeOpcUa/opcua-asyncio <opcua-asyncio@noreply.github.com>
Cc: pfetrial <patrick.frey37@sfr.fr>; Author <author@noreply.github.com>
Betreff: Re: [FreeOpcUa/opcua-asyncio] Historizing (#375)

 

if you use the default historymanager then yes

**Reply by AndreasHeine** [reply]
with sqlite is both possible depends on your code!

file:
`server.iserver.history_manager.set_storage(HistorySQLite("my_datavalue_history.sql"))`
ram:
`server.iserver.history_manager.set_storage(HistorySQLite(":memory:"))`

there is also a HistoryDict!
